src/utilities/createRouterFind.ts

```diff
@@ -1,24 +1,40 @@
 import { ResolvedRoute } from '@/types/resolved'
 import { RouterRoutes } from '@/types/routerRoute'
-import { RouteWithParams } from '@/types/routeWithParams'
-import { RouterResolve } from '@/utilities/createRouterResolve'
+import { RoutesKey } from '@/types/routesMap'
+import { RouteKeysThatHaveOptionalParams, RouteKeysThatHaveRequireParams, RouteParamsByName } from '@/types/routeWithParams'
+import { Url } from '@/types/url'
+import { createRouterResolve } from '@/utilities/createRouterResolve'
 import { getResolvedRouteForUrl } from '@/utilities/getResolvedRouteForUrl'
 
 export type RouterFind<
   TRoutes extends RouterRoutes
-> = <
-  TRoutePath extends string
->(source: string | RouteWithParams<TRoutes, TRoutePath>) => ResolvedRoute | undefined
-
-type CreateRouterFindContext<T extends RouterRoutes> = {
-  routes: T,
-  resolve: RouterResolve<T>,
+> = {
+  <TSource extends RouteKeysThatHaveRequireParams<TRoutes>>(source: TSource, params: RouteParamsByName<TRoutes, TSource>): ResolvedRoute | undefined,
+  <TSource extends RouteKeysThatHaveOptionalParams<TRoutes>>(source: TSource, params?: RouteParamsByName<TRoutes, TSource>): ResolvedRoute | undefined,
+  (source: Url): ResolvedRoute | undefined,
 }
 
-export function createRouterFind<const T extends RouterRoutes>({ routes, resolve }: CreateRouterFindContext<T>): RouterFind<T> {
-  return (source) => {
-    const url = resolve(source)
+export function createRouterFind<const TRoutes extends RouterRoutes>(routes: TRoutes): RouterFind<TRoutes> {
```

Comment:
Same thought here (and for all the rest of the methods declared like this).

Reply:
for RouterResolve and RouterReplace these functions actually have the correct types, not sure why those work well but find/replace inside of createRouter require a bit more coercing